src/main/java/com/android/volley/toolbox/BasicNetwork.java

```diff
@@ -313,16 +313,43 @@ protected void logError(String what, String url, long start) {
         return result;
     }
 
-    // TODO(#21): Pass all headers to NetworkResponse so clients can access all the headers with
-    // the same keys (e.g. for SetCookie headers). Just need to ensure case-insensitivity.
-    private static Map<String, String> convertHeaders(Map<String, List<String>> headers) {
-        Map<String, String> result = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
-        for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
-            if (entry.getKey() != null) {
-                List<String> values = entry.getValue();
-                result.put(entry.getKey(), values.get(0));
+    /**
+     * Combine cache headers with network response headers for an HTTP 304 response.
+     *
+     * <p>An HTTP 304 response does not have all header fields. We have to use the header fields
+     * from the cache entry plus the new ones from the response. See also:
+     * http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.5
+     *
+     * @param responseHeaders Headers from the network response.
+     * @param entry The cached response.
+     * @return The combined list of headers.
+     */
+    private static List<Header> combineHeaders(List<Header> responseHeaders, Entry entry) {
+        // First, create a case-insensitive set of header names from the network
+        // response.
+        Set<String> headerNamesFromNetworkResponse =
+                new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+        for (Header header : responseHeaders) {
+            headerNamesFromNetworkResponse.add(header.getName());
+        }
+
+        // Second, add headers from the cache entry to the network response as long as
+        // they didn't appear in the network response, which should take precedence.
+        List<Header> combinedHeaders = new ArrayList<>(responseHeaders);
+        if (entry.allResponseHeaders != null) {
```

Comment:
I think if you also check for emptiness here you'll save an iterator allocation (when the collection is empty).

Reply:
Thanks, done. (Though in practice I think it'd be pretty surprising for an HTTP response to not include any response headers).

Reply:
Agreed.